src/main/kotlin/com/goodchoice/domain/review/persistence/ReviewRepository.kt

```diff
@@ -0,0 +1,165 @@
+import com.goodchoice.domain.common.model.Reference
+import com.goodchoice.domain.review.model.*
+import com.goodchoice.domain.subject.model.Mark
+import com.goodchoice.infra.common.now
+import com.goodchoice.infra.persistence.read
+import org.jooq.DSLContext
+import java.time.Clock
+import java.util.*
+
+interface ReviewRepository {
+    fun create(
+        title: String,
+        author: Reference,
+        subject: Reference,
+        advantages: List<String>,
+        disadvantages: List<String>,
+        mark: Mark,
+        body: ReviewBodyCreationRequest
+    ): Reference
+
+    fun vote(reviewId: UUID, issuerId: UUID, voteType: VoteType)
+    fun getVotesByReviewIdOrNull(reviewId: UUID, issuerId: UUID): ReviewVotes?
+    fun removeVote(reviewId: UUID, issuerId: UUID)
+    fun getAllBySubject(subjectId: UUID, mark: Mark?, issuerId: UUID?, pageRequest: PageRequest): Page<Review>
+}
+
+class ReviewJooqRepository(
+    private val db: DSLContext,
+    private val clock: Clock,
+    private val objectMapper: ObjectMapper
+) : ReviewRepository {
+    override fun create(
+        title: String,
+        author: Reference,
+        subject: Reference,
+        advantages: List<String>,
+        disadvantages: List<String>,
+        mark: Mark,
+        body: ReviewBodyCreationRequest
+    ): Reference {
+        val reviewId = UUID.randomUUID()
+        val reviewBodyId = UUID.randomUUID()
+
+        db.insertInto(REVIEW)
+            .set(REVIEW.ID, reviewId)
+            .set(REVIEW.TITLE, title)
+            .set(REVIEW.SUBJECT_ID, subject.id)
+            .set(REVIEW.MARK, mark.value)
+            .set(REVIEW.REVIEWER_ID, author.id)
+            .set(REVIEW.IS_SHOWN, true)
+            .execute()
+
+        db.insertInto(REVIEW_BODY)
+            .set(REVIEW_BODY.ID, reviewBodyId)
+            .set(REVIEW_BODY.CONTENT, body.content)
+            .set(REVIEW_BODY.CREATED_TIMESTAMP, clock.now())
+            .set(REVIEW_BODY.REVIEW_ID, reviewId)
+            .execute()
+
+        var ordering = 0
+        db.insertInto(
+            REVIEW_POINT,
+            REVIEW_POINT.CONTENT,
+            REVIEW_POINT.ORDERING,
+            REVIEW_POINT.REVIEW_ID,
+            REVIEW_POINT.TYPE
+        ).apply { advantages.forEach { values(it, ordering++, reviewId, ADVANTAGE) } }
+            .apply { disadvantages.forEach { values(it, ordering++, reviewId, DISADVANTAGE) } }
+            .execute()
+
+        return Reference(reviewId)
+    }
+
+    override fun vote(reviewId: UUID, issuerId: UUID, voteType: VoteType) {
+        db.insertInto(REVIEW_VOTE, REVIEW_VOTE.REVIEWER_ID, REVIEW_VOTE.REVIEW_ID, REVIEW_VOTE.TYPE)
+            .values(issuerId, reviewId, voteType.asJooqVoteType())
+            .onDuplicateKeyUpdate()
+            .set(REVIEW_VOTE.REVIEWER_ID, issuerId)
+            .set(REVIEW_VOTE.REVIEW_ID, reviewId)
+            .set(REVIEW_VOTE.TYPE, voteType.asJooqVoteType())
+            .execute()
```

Comment:
We should use either `db.insertInto(TABLE).set(FIELD, value)` or `db.insertInto(TABLE, FIELD).values(value)`

Reply:
onDuplicateKeyUpdate() requires both values and set, see https://www.jooq.org/doc/3.0/manual/sql-building/sql-statements/insert-statement/insert-on-duplicate-key/

Reply:
Ok, sorry